### PR TITLE
#1700 Abandon all the services if an error occurs while creating a workspace

### PR DESCRIPTION
**Pull request by wu87988622**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #1700 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/api/workerApi.js [hunk 1]
```diff
@@ -104,3 +104,19 @@ export const addNodeToWorker = async params => {
     handleError(err);
   }
 };
+
+export const deleteWorker = async name => {
```
Comment (saivirtue) [inline ohara-manager/client/src/api/workerApi.js]: Seems this const not used?

Comment (wu87988622) [reply ohara-manager/client/src/api/workerApi.js]: ok I will remove this.
thanks

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 2]
```diff
@@ -123,6 +130,82 @@ const WorkerNewModal = props => {
     return value;
   };
 
+  const waitDeleteService = async params => {
+    const { service, name } = params;
+    let { retryCount } = params;
+    const maxRetry = 10;
+    switch (service) {
+      case 'zk':
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: I would prefer the full name of the service like broker or zookeeper :)

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 3]
```diff
@@ -48,6 +48,8 @@ const WorkerNewModal = props => {
   const [jars, setJars] = useState([]);
   const [isLoading, setIsLoading] = useState(false);
   const [errorMessenge, setErrorMessenge] = useState('');
+  let workingService = [];
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: `workingServices`

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 4]
```diff
@@ -56,6 +58,9 @@ const WorkerNewModal = props => {
 
     const isSuccess = get(res, 'data.isSuccess', false);
     if (isSuccess) {
+      const newPlugins = [...plugins];
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: Doesn't `newPlugins = [...plugins, res]` work in this senario?

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 5]
```diff
@@ -123,6 +130,82 @@ const WorkerNewModal = props => {
...
+        retryCount++;
+        await waitDeleteService({ service, name, retryCount });
+        return;
+      default:
+        return;
+    }
+  };
+
+  const saveService = params => {
+    const newSave = [...workingService];
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: What is `newSave`...? NewService?

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 6]
```diff
@@ -123,6 +130,82 @@ const WorkerNewModal = props => {
...
+    newSave.push(params);
+    workingService = newSave;
+  };
+
+  const deleteAllSerice = async () => {
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: `deleteAllservices`

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 7]
```diff
@@ -123,6 +130,109 @@ const WorkerNewModal = props => {
...
+    plugins.forEach(plugin => {
+      const { name, group } = plugin.data.result;
+      jarApi.deleteJar({ name: name, workerClusterName: group });
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: Object shorthand: `jarApi.deleteJar({name, workerClusterName: group})`

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 8]
```diff
@@ -123,6 +130,109 @@ const WorkerNewModal = props => {
...
+    });
+    const wks = workingService.filter(working => working.service === 'wk');
+    const bks = workingService.filter(working => working.service === 'bk');
+    const zks = workingService.filter(working => working.service === 'zk');
+
+    await Promise.all(
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: Why not just

```js
await workerApi.deleteWorker(`${wk.name}`);
await waitDeleteService({
  service: 'wk',
  name: wk.name,
  retryCount: 0,
});
```

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js [hunk 9]
```diff
@@ -123,6 +130,109 @@ const WorkerNewModal = props => {
     return value;
   };
 
+  const waitDeleteService = async params => {
+    const { service, name } = params;
+    let { retryCount } = params;
```
Comment (konekoya) [inline ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: You can use the default param here for the retryCount

```js
let { retryCount = 0 } = params

// So you don't have to call it like this:
waitDeleteService({ retryCount = 0 })
```

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js (hunk 4 above)
Comment (wu87988622) [reply ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: magic!!

File: ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js (hunk 7 above)
Comment (wu87988622) [reply ohara-manager/client/src/components/pages/WorkspacesPage/WorkspacesNewModal/WorkspacesNewModal.js]: magic~~

**Comment by konekoya** [comment]
retry manager-api

**Comment by konekoya** [comment]
retry manager-it

**Comment by konekoya** [comment]
@wu87988622 if the failing tests are not related to this PR then I think it's ready to be merged

**Comment by konekoya** [comment]
retry manager-it

**Pull request merged**